CAAM driver minor fixes

clementfaure opened the pull request:
Hello,

Here are some patches for the CAAM driver. Most of these patches fix minor issues in the CAAM.
However, the patch fixing the issue I found last week in xtest is the following : `drivers: caam: fix missing output additional length`

Thanks,
Clement

jforissier reviewed (commented):
One general comment: when a commit is a fix for a previous commit that is already part of a released version of OP-TEE, it is nice to add a `Fixes:` tag citing the commit that is being fixed. The reason is to make it easier for downstream projects to select and pick patches based on what they have in their tree.

Consider for example commit 1d88c0c03f3bed5776739e459a598a011baa5207 in release 3.14.0 which fixes 59ac3801b756 introduced in 3.13.0.

Now, sometimes we add `Fixes:` tags for bugs introduced in the same release (probably less useful) or sometimes we forget to add them... but I think we should try to do it more consistently. `git blame <file>`, `git show <commit>`, `git tag --contains <commit>` are helpful commands.

clementfaure commented on core/drivers/crypto/caam/hal/common/hal_ctrl.c:
```diff
@@ -73,9 +73,9 @@ uint8_t caam_hal_ctrl_hash_limit(vaddr_t baseaddr)
+	uint32_t val = io_caam_read32(baseaddr + CTPR_LS);
 
-	return GET_CAAMVID_MS_MAJ_REV(val) >= 3;
+	return !!GET_CTPR_LS_SPLIT_KEY(val);
```
I will remove it

clementfaure commented on core/drivers/crypto/crypto_api/acipher/ecc.c:
```diff
@@ -98,16 +98,19 @@ static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
 		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
+	key_size = get_ecc_key_size_bytes(key->curve) * 8;
```
Sorry about that, that's the result of squashing two commits .. I will split this commits into two.

clementfaure commented:
Thanks for your feedback @jforissier 
It's true that we tend to forget about adding `Fixes: ` to our commit description. We will try on our side to stick to this good practice.
I added `Fixes: ` to these commits. Most of them fix initial CAAM driver patches. I also addressed your two comments. Let me know if that works for you.

jforissier commented on core/drivers/crypto/crypto_api/acipher/ecc.c:
```diff
@@ -98,16 +98,19 @@ static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
+		ret = ecc->gen_keypair(key, key_size);
 
 	CRYPTO_TRACE("ECC Keypair (%zu bits) generate ret = 0x%" PRIx32,
 		     size_bits, ret);
```
s/size_bits/key_size/?

clementfaure replied:
Oops ... fixed

jforissier commented on core/drivers/crypto/caam/hal/common/hal_ctrl.c:
```diff
@@ -75,7 +75,7 @@ bool caam_hal_ctrl_splitkey_support(vaddr_t baseaddr)
 	uint32_t val = io_caam_read32(baseaddr + CTPR_LS);
 
-	return !!GET_CTPR_LS_SPLIT_KEY(val);
+	return !GET_CTPR_LS_SPLIT_KEY(val);
```
Oops! You left one `!`.

clementfaure replied:
Oops ... fixed

etienne-lms replied:
note it's been fixed in the wrong commit.

jforissier reviewed (commented):
With the two things below addressed, LGTM.
`Acked-by: Jerome Forissier <jerome@forissier.org>`

clementfaure commented:
Force-push:
- Squash review commits
- Address the last two comments
- Applied tags

etienne-lms commented on core/drivers/crypto/crypto_api/acipher/ecc.c:
```diff
@@ -98,19 +98,22 @@ static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
 {
 	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
 	struct drvcrypt_ecc *ecc = NULL;
+	size_t key_size = 0;
```
for consistency, prefer rename to `key_size_bits`.

clementfaure replied:
I agree

etienne-lms commented on core/drivers/crypto/caam/hal/common/hal_ctrl.c:
```diff
@@ -75,7 +75,7 @@ bool caam_hal_ctrl_splitkey_support(vaddr_t baseaddr)
 	uint32_t val = io_caam_read32(baseaddr + CTPR_LS);
 
-	return !GET_CTPR_LS_SPLIT_KEY(val);
+	return GET_CTPR_LS_SPLIT_KEY(val);
```
should belong to another commit.

clementfaure replied:
My bad, I will fix it

etienne-lms reviewed (commented):
comments for commit "drivers: crypto: fix ECC key size bytes to bits conversion".

etienne-lms commented on core/drivers/crypto/caam/utils/utils_dmaobj.c:
```diff
@@ -915,8 +921,14 @@ size_t caam_dmaobj_copy_ltrim_to_orig(struct caamdmaobj *obj)
+	 * (obj->orig.length) and cannot exceed the data processed by the
+	 * CAAM (obj->sgtbuf.length).
+	 */
+	dst_rlen = MIN(obj->orig.length, obj->sgtbuf.length);
```
this changes `dst_rlen` value. I think test at line 938 is corrupted by this change.

clementfaure replied:
What do you mean by corrupted? To me, this test still looks valid

etienne-lms replied:
With previous implementation `dst_rlen` was 0 when  `off >= obj->orig.length`. Here it likely always gets non 0 value. It could cause trouble IMO.

clementfaure replied:
You are right, the change might have an impact on the following `if (!dst_rlen) {` statement.
In the end, this change in the function `caam_dmaobj_copy_ltrim_to_orig()` is not needed. This function is only used in the RSA crypto operation and the `obj->orig.length` cannot be smaller than `obj->sgtbuf.length` because of the SGT allocation for the RSA. Thus, the modification done is not needed.

clementfaure replied:
@etienne-lms Gentle reminder!

etienne-lms reviewed (commented):
comment for commit "drivers: caam: dma: fix copy data size"

clementfaure commented:
Force-push:
* Split commit `drivers: crypto: fix ECC key size bytes to bits conversion` to fix the CAAM split key in the appropriate commit

ruchi393 commented:
`Acked-by: Ruchika Gupta <ruchika.gupta@linaro.org>`

clementfaure commented:
Force-push:
- Applied @ruchi393 tags
- The CI failed for some reasons so I rebased the PR.

@etienne-lms Do you have any other comments/questions?

clementfaure commented on core/drivers/crypto/caam/hal/common/registers/rng_regs.h:
```diff
@@ -22,27 +22,88 @@
 #define TRNG_SDCTL_ENT_DLY(val)		SHIFT_U32(((val) & 0xFFFF), 16)
+#define BM_TRNG_SDCTL_SAMP_SIZE		SHIFT_U32(0xFFFF, 0)
+#define GET_TRNG_SDCTL_SAMP_SIZE(val)	((val) & (BM_TRNG_SDCTL_SAMP_SIZE))
+#define TRNG_SDCTL_SAMP_SIZE(val)	SHIFT_U32(((val) & (0xFFFF)), 0)
```
Thanks for noticing it! I will remove these values.
I will also reformat `TRNG_SDCTL_SAMP_SIZE`

clementfaure commented:
Force-push:
* Squash review commits
* Tag applied (@etienne-lms) Thanks!

clementfaure commented:
Force-push:
* Rebase
